File: 코드/ch8/clip1/boj3078/main.py
import sys
from collections import deque

n, k = map(int, sys.stdin.readline().split())
students = [len(sys.stdin.readline().rstrip()) for _ in range(n)]

friends = [0] * 21
queue = deque()
sum_ = 0

for i in range(n):
    next_ = students[i]
    sum_ += friends[next_]
    
    queue.append(next_)
    friends[next_] += 1
    if len(queue) > k:
        out = queue.popleft()
        friends[out] -= 1

sys.stdout.write(str(sum_) + '\n')

File: 코드/ch8/clip1/boj3078/Main.java
import java.util.LinkedList;
import java.util.Queue;
import java.util.Scanner;

public class Main {
    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        int n = sc.nextInt();
        int k = sc.nextInt();
        int[] students = new int[n];
        long sum = 0;
        for(int i=0; i<n; i++) {
            String str = sc.next();
            students[i] = str.length();
        }
        // 등수가 k보다 작거나 같은 좋은친구를 카운팅할 빈도수
        int[] friends = new int[21];
        Queue<Integer> queue = new LinkedList<>();
        for(int i=0; i<n; i++) {
            int next = students[i];
            sum += friends[next];

            queue.add(next);
            friends[next]++;
            if(queue.size() > k) {
                int out = queue.poll();
                friends[out]--;
            }
        }
        System.out.println(sum);
    }
}

